Adding Test Cases for Sv32 VM

Ammar-10xe opened the pull request:
## [Test for  SATP Access Permission ](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/d375502c80824e21d06ac451146d8b09e11465b2)

Please review this test case and the overall implementation to ensure they align with the arch test specifications.This PR introduces a custom macros.h  header file also, that is used in this test case and will be used in the other upcoming tests also

**Summary:**

This pull request adds a test case to verify the access control behavior of the satp (Supervisor Address Translation and Protection) register. The goal of this test is to ensure that the satp register is only accessible in Machine (M) and Supervisor (S) privilege modes, while generating an illegal instruction exception when accessed in lower privilege modes.

Ammar-10xe committed:
initial commit

Ammar-10xe committed:
custom macros file updated

Ammar-10xe committed:
test name changed

Ammar-10xe committed:
Invalid PTE Permission test case for Level0 and level1 added

Ammar-10xe commented:
## [Test for In-Valid Permission of PTE](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/77459aeb22d16fccfb185651f3c6dc6f9b6560ab)

I've added a new test case to cover scenarios involving invalid permissions of Page Table Entries (PTEs). When a PTE does not have valid permission (pte.V=0), attempting to access it will result in a page fault exception of the corresponding access type.

The test cases are specifically designed for situations when satp.mode=sv32 and the PTE has (r,w,x) PMP (Physical Memory Protection) permissions. The following tests have been implemented and executed in both supervisor and user privilege modes for both level0 and level1 PTEs:

1. Set PTE.V = 0 and test read access.
2. Set PTE.V = 0 and test write access.
3. Set PTE.V = 0 and test execute access.

Please review the changes and let me know if any further adjustments or additions are needed. Your feedback is highly appreciated!

#### Files and their purpose:

|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_invalid_PTE_01.S** 	|   Supervisor	|     1     	|        Read Access  	|
| **vm_invalid_PTE_02.S** 	|   Supervisor	|     1     	|    	 Write Access     	|   
| **vm_invalid_PTE_03.S** 	|    Supervisor 	|     1     	|      	 Execute Access      |   
| **vm_invalid_PTE_04.S** 	|    User	|     1     	|        Read Access  	|
| **vm_invalid_PTE_05.S** 	|    User 	|     1     	|        Write Access     	|      
| **vm_invalid_PTE_06.S** 	|    User 	|     1     	|        Execute Access      |    
| **vm_invalid_PTE_07.S** 	|    Supervisor 	|     0    	|        Read Access  	|
| **vm_invalid_PTE_08.S** 	|    Supervisor 	|     0     	|        Write Access       	|  
| **vm_invalid_PTE_09.S** 	|    Supervisor 	|     0     	|        Execute Access      |     
| **vm_invalid_PTE_10.S** 	|    User 	|     0    	|        Read Access  	|
| **vm_invalid_PTE_11.S** 	|    User 	|      0   	|        Write Access       	|  
| **vm_invalid_PTE_12.S** 	|    User 	|     0     	|        Execute Access      |

Commit:
Test for Dirty bit implementation added
If implementation does not sets the pte.D when PTE is written, and PTE has pte.D=0,
then attempting to store on that PTE would raise Store/AMO page fault exception.

MuhammadHammad001 commented:
## Test for Dirty Bit Implementation

#### Feature Description:
If implementation does not sets the pte.D when PTE is written, and PTE has pte.D=0, then attempting to store on that PTE would raise **Store/AMO page fault exception**.

#### Macro for the test:
**TEST_DIRTY_BIT(MODE, LEVEL, R, W, X)**
_Note::_ R,W,X &rarr; permission bits for PTE

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Read`** 	| **`Write`** 	| **`Execute`** 	|
|:-----------------:	|:----------:	|:---------:	|:--------:	|:---------:	|:-----------:	|
| **vmDirtyBit-01** 	| supervisor 	|     1     	|     1    	|     1     	|      1      	|
| **vmDirtyBit-02** 	|    user	|     1     	|     1    	|     1     	|      1      	|
| **vmDirtyBit-03** 	| supervisor 	|     0     	|     1    	|     1     	|      1      	|
| **vmDirtyBit-04** 	|    user 	|     0     	|     1    	|     1     	|      1      	|

HAMZA-AFZAL404 committed:
U_bitunset tests case added

HAMZA-AFZAL404 commented:
## [Test for U Bit Implementation](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/1c28c0af30a1162779c40676262b1e59ad297668)

#### Feature Description:
If implementation does not sets the pte.U **Page fault exceptions** should be raised according to {x,r,w} access type.

**TEST_U_BIT(MODE, LEVEL, R,W,X) **
_Note::_ R,W,X Means the which i am testing for

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_PTE_U_unset-01** 	| user 	|     1     	|        Execute Access    	|
| **vm_PTE_U_unset-02** 	|    user	|     1     	|    	     Read Access      	|   
| **vm_PTE_U_unset-03** 	| user 	|     1     	|      	     Write Access        	|   
| **vm_PTE_U_unset-04** 	|    user 	|     0     	|          Execute Access    	| 
| **vm_PTE_U_unset-05** 	|    user 	|     0     	|         Read Access     	|    
| **vm_PTE_U_unset-06** 	|    user 	|     0     	|          Write Access       	|

Ammar-10xe committed:
RWX access on S mode pages in S-mode

Ammar-10xe commented:
## [Test for RWX access on S-mode pages in S-mode](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/2a1d1b6b9bc886d6e1a036ceef76463fac6bd348)

#### Feature Description:
If PTE belongs to supervisor mode i.e. its U permission bit is clear (pte.u = 0), then accessing that PTE in supervisor mode should be successful if the corresponding (r,w,x) permission of PTE is granted. Otherwise raise page fault exception of the corresponding access type.

When satp.mode=sv32, PTE has (r,w,x) PMP permissions, PTE has non-reserved RWX encoding, pte.u=0 and pte.v=1, then test the following in supervisor privilege mode for level 0 and level 1 PTE.

- Test the read access for both pte.r=1 and for pte.r=0
- Test the write access for both pte.w=1 and for pte.w=0
- Test the execute access for both pte.x=1 and for pte.x=0

For testing at level0, set all the pte perimssions at level1 to 0 except pte.v, so that level1 points to level0.



#### Files and their purpose:

|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_RWX_Access_S_01.S** 	|   Supervisor	|     1     	|        Read Access  (PTE.r = 1) 	|
| **vm_RWX_Access_S_02.S** 	|   Supervisor	|     1     	|    	 Write Access  (PTE.w = 1)    	|   
| **vm_RWX_Access_S_03.S** 	|    Supervisor 	|     1     	|      	 Execute Access (PTE.x= 1)   |   
| **vm_RWX_Access_S_04.S** 	|    Supervisor 	|     0     	|        Read Access  (PTE.r = 1)  	|
| **vm_RWX_Access_S_05.S** 	|    Supervisor 	|     0     	|        Write Access   (PTE.w = 1)   	|      
| **vm_RWX_Access_S_06.S** 	|    Supervisor 	|     0     	|        Execute Access  (PTE.x = 1) |    
| **vm_RWX_Access_S_07.S** 	|    Supervisor 	|     1    	|        Read Access  (PTE.r = 0) 	|
| **vm_RWX_Access_S_08.S** 	|    Supervisor 	|     1     	|        Write Access   (PTE.w = 0)    |  
| **vm_RWX_Access_S_09.S** 	|    Supervisor 	|     1     	|        Execute Access  (PTE.x = 0) |     
| **vm_RWX_Access_S_10.S** 	|    Supervisor 	|     0     	|        Read Access  (PTE.r = 0) 	|
| **vm_RWX_Access_S_11.S** 	|    Supervisor 	|     0     	|        Write Access   (PTE.w = 0)    |  
| **vm_RWX_Access_S_12.S** 	|    Supervisor 	|     0     	|        Execute Access  (PTE.x = 0) |

HAMZA-AFZAL404 committed:
U_bit set tests added and stisfied

HAMZA-AFZAL404 committed:
Merge remote-tracking branch 'refs/remotes/origin/sv32-arch-test' into sv32-arch-test

HAMZA-AFZAL404 commented:
## [Test for U Bit Implementation](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/e32728d1db7c127d1760aa920eae791329fde86b)

#### Feature Description:
If U permission bit is set (pte.u = 1), then accessing that PTE in user mode should be successful if the corresponding (r,w,x) permission of PTE is granted. Otherwise raise page fault exception of the corresponding access type.

**TEST_U_BIT(MODE, LEVEL, R,W,X) **
_Note::_ R,W,X Means the which i am testing for

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_PTE_U_unset-01** 	| user 	|     1     	|        Execute Access  with X=1  	|
| **vm_PTE_U_unset-02** 	|    user	|     1     	|    	     Execute Access       	|   
| **vm_PTE_U_unset-03** 	| user 	|     1     	|      	     Read Access         	|   
| **vm_PTE_U_unset-04** 	|    user 	|     1     	|          Write Access   	| 
| **vm_PTE_U_unset-05** 	|    user 	|     0     	|         Execute Access  with X=1     	|    
| **vm_PTE_U_unset-06** 	|    user 	|     0     	|          Execute Access       	|  
| **vm_PTE_U_unset-07** 	|    user 	|     0     	|         Read Access     	|    
| **vm_PTE_U_unset-08** 	|    user 	|     0     	|          Write Access       	|

HAMZA-AFZAL404 committed:
bare mode tests added

HAMZA-AFZAL404 commented:
## [Test for Satp Bare Mode](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/29f15621d20e09e8084cb05cb6bf716dbe9e649a)

#### Feature Description:
Show writing {zero, non-zero} value to satp when mode=bare the behavior follows the design implementation.

**TEST_Bare_Mode(MODE, LEVEL, Va and Pa) **
_Note::_ R,W,X Means the which i am testing for

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Va and Pa`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_Bare_mode-01** 	| user 	|     1     	|        Equal 	|
| **vm_Bare_mode-02** 	|    user	|     1     	|    	     Not Equal      	|   
| **vm_Bare_mode-03** 	| user 	|     0     	|      	     Equal         	|   
| **vm_Bare_mode-04** 	|    user 	|     0     	|          Not Equal   	|

UmerShahidengr commented on cva6/tests/custom/sv32/arch_tests/custom_macros.h:
```diff
@@ -0,0 +1,201 @@
+#include "encoding.h"
+
```
We need to add these macros to [test_macros.h](https://github.com/riscv-non-isa/riscv-arch-test/blob/main/riscv-test-suite/env/test_macros.h). This file is responsible of keeping the definitions of all macros which have been used by the tests, and is also used while calculating coverage.

Ammar-10xe replied:
Thank you, @UmerShahidengr   for your thorough review! based on your feedback, I'll be updating the custom macros implementation using `#ifdef` statements. This adjustment will ensure that if the macros aren't found in the [test_macros.h](https://github.com/riscv-non-isa/riscv-arch-test/blob/main/riscv-test-suite/env/test_macros.h) file, they will be sourced from the [custom_macros.h](https://github.com/10x-Engineers/core-v-verif/pull/9/files/29f15621d20e09e8084cb05cb6bf716dbe9e649a#diff-9e64449c0d65edba4e6878f0077cd3ae68804129ee20c2ba788080b2c4243b4d)  file as intended.

UmerShahidengr commented on cva6/tests/custom/sv32/arch_tests/vm_Bare_mode_01.S:
```diff
@@ -0,0 +1,133 @@
+#                                                                                               #
+################################################################################################# 
+#include "model_test.h"
+#include "custom_macros1.h"
```
This line will be removed from all tests, as custom_macros will be merged in test_macors

Ammar-10xe replied:
Done

UmerShahidengr commented on cva6/tests/custom/sv32/arch_tests/vm_Bare_mode_01.S:
```diff
@@ -0,0 +1,133 @@
+	add t2, t1, t0
+	SREG t2, code_bgn_off+1*sv_area_sz(sp)
+    SREG t2, code_bgn_off+2*sv_area_sz(sp)
+
```
This test is returning nothing in signature region, since arch_tests are signature based tests, so it will always pass as there is nothing in the signature file to compare against DUT and ref.
RVTEST_SIGUPD() macro is used to write data into signature, this macro has to be used to export data to be compared. This macro has some bug as it always save physical address of signature area into based register (x13 in this case), so you have to setup a PTE for that base register address and do the address translation if you want to save something in S-mode. One way of doing it can be:

```
	li a0, 0x80006000   // Physical address of signature area
	li a1, (PTE_V | PTE_R | PTE_A | PTE_W | PTE_X | PTE_D)
	PTE_SETUP_RV32(a0, a1, t0, t1, 0x90006000, 0) // PTE Setup for signature area address



	LI (t0, va)
	LI (t1, pa)
	sub t0, t0, t1 // (VA-PA) Note: VA > PA 
	add x13, x13, t0	// Translation of Signature reg
	
	RVTEST_SIGUPD(x13,a4)  // Now x13 will have VA and it will do the signature update.

```

Ammar-10xe replied:
Thank you, @UmerShahidengr , I appreciate your thorough explanation of the issue and the usage of the `RVTEST_SIGUPD() `macro. I'll be updating all the test cases to ensure proper dumping of signatures into the signature file.

UmerShahidengr commented on cva6/tests/custom/sv32/arch_tests/vm_Bare_mode_02.S:
Template of these tests are fine, and I am assuming that you have verified its correctness via Sail logs, its just about the updates of exporting some relevant data into signature file is missing. Otherwise tests are fine.

Ammar-10xe replied:
I'm glad you found the test template satisfactory. I'd like to confirm that I've diligently cross-referenced the Sail logs and rigorously reviewed each test case. Rest assured, the purpose of each test case has been thoroughly met. Regarding the export of relevant data into the signature file, I'll be working on updating the code to address this. I'll ensure to deliver the changes promptly to resolve this matter.

Ammar-10xe committed:
RWX access on U-mode pages in S-mode with s/mstatus.SUM set

Ammar-10xe commented:
## [Test for RWX access on U-mode pages in S-mode with s/mstatus.SUM set](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/8cb5238045357aa9534adcb3e567ef0a745af31c)

If PTE belongs to user mode i.e. its U permission bit is set (pte.u = 1) and m/sstatus.SUM = 1, then RW access to that PTE in supervisor mode would be successful but eXecute access would raise instruction page fault exception in s-mode.
When satp.mode=sv32, PTE has (r,w,x) PMP permissions, PTE has non-reserved RWX encoding, and pte.v=1, then test the following in supervisor mode for level0 and level1 PTE.

- Set pte.r=1 & pte.u=1 & s/mstatus.SUM = 1 and test the read acces.
- Set pte.w=1 & pte.u=1 & s/mstatus.SUM = 1 and test the write access.
- Set pte.x=1 & pte.u=1 & s/mstatus.SUM = 1 and test the execute access.

For testing at level0, set all the pte perimssions at level1 to 0 except pte.v, so that level1 points to level0.

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_RWX_Access_U_01.S** 	|   Supervisor	|     1     	|   Read Access  	|
| **vm_RWX_Access_U_02.S** 	|   Supervisor	|     1     	|   Write  Access      |   
| **vm_RWX_Access_U_03.S** 	|   Supervisor 	|     1     	|   Execute Access  |   
| **vm_RWX_Access_U_04.S** 	|   Supervisor 	|     0     	|   Read Access  	|
| **vm_RWX_Access_U_05.S** 	|   Supervisor 	|     0     	|   Write  Access       | 
| **vm_RWX_Access_U_06.S** 	|   Supervisor 	|     0     	|   Execute Access   |

Ammar-10xe committed:
RWX access Sum set and unset test cases added

Ammar-10xe committed:
RWX access on U-mode pages in S-mode with s/mstatus.SUM unset

Ammar-10xe commented:
## [Test for RWX access on U-mode pages in S-mode with s/mstatus.SUM unset](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/78dd4940c8fb439e96d6ce54eed5c43dbf148910)

If PTE belongs to user mode i.e. its U permission bit is set (pte.u = 1) and m/sstatus.SUM = 0, then accessing that PTE in supervisor mode would raise page fault exception of the corresponding access type. 
When satp.mode=sv32, PTE has (r,w,x) PMP permissions, PTE has non-reserved RWX encoding, and pte.v=1, then test the following in supervisor mode for level 0 and level 1 PTE.
- Set pte.u=1 & s/mstatus.SUM = 0 and test the read acces.
- Set pte.u=1 & s/mstatus.SUM = 0 and test the write access.
- Set pte.u=1 & s/mstatus.SUM = 0 and test the execute access.

For testing at level0, set all the pte perimssions at level1 to 0 except pte.v, so that level1 points to level0.

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_SUM_unset_01.S** 	|   Supervisor	|     1     	|   Read Access  	|
| **vm_SUM_unset_02.S** 	|   Supervisor	|     1     	|   Write  Access      |   
| **vm_SUM_unset_03.S** 	|   Supervisor 	|     1     	|   Execute Access  |   
| **vm_SUM_unset_04.S** 	|   Supervisor 	|     0     	|   Read Access  	|
| **vm_SUM_unset_05.S** 	|   Supervisor 	|     0     	|   Write  Access       | 
| **vm_SUM_unset_06.S** 	|   Supervisor 	|     0     	|   Execute Access   |

Ammar-10xe committed:
Misaligned Superpage test case added

Ammar-10xe commented:
## [Test for Misaligned Superpage](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/f3689c7815016536722edbc9b531ef83aac33414)

If PTE at level1 is leaf PTE (superpage) and its pte.ppn[0]!=0, then it is a misaligned superpage and accessing that PTE would raise page fault exception of the corresponding access type.
When satp.mode=sv32, PTE has (r,w,x) PMP permissions, PTE has non-reserved RWX encodings, (pte.r | pte.x)=1, and pte.v=1, then test the following in supervisor and user privilege mode for level1 PTE.
-set pte.ppn[0]!=0 and test for read, write and execute access.

**Set pte.U=0 when test in Supervisor mode and Set pte.U=1 when testing in user mode.**

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_misaligned_superpage01.S** 	|   Supervisor	|     1     	|   Read Access  	|
| **vm_misaligned_superpage02.S** 	|   Supervisor	|     1     	|   Write  Access      |   
| **vm_misaligned_superpage03.S** 	|   Supervisor 	|     1     	|   Execute Access  |   
| **vm_misaligned_superpage04.S** 	|   User 	|     1     	|   Read Access  	|
| **vm_misaligned_superpage05.S** 	|   User 	|     1     	|   Write  Access       | 
| **vm_misaligned_superpage06.S** 	|   User 	|     1     	|   Execute Access   |

Ammar-10xe committed:
bare mode test name chnaged

Ammar-10xe committed:
tests named changed to small letters

Ammar-10xe committed:
test custom macros.h file updated

Ammar-10xe committed:
test_custom_macros.h file updated

Ammar-10xe committed:
test satp_asid.S added

Ammar-10xe commented:
## [Test for SATP ASIDLEN ](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/8dd213384be0ab4e2926db17f42d2b4104b703ec)

The purpose of this test to determine by writing one to every bit position in the ASID field, then reading back the value in satp to see which bit positions in the ASID field hold a one. Show that ASIDLEN is equal to the expected ASIDLEN.

**NOTE:**
ASIDLEN is the number of ASID bits implemented. MAXASID bits for sv32 is 9

Ammar-10xe committed:
PMP permissions on Physical Address

Ammar-10xe commented:
## [Test for PMP permissions on Physical Address](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/140c090d7914275e25446ac2715f6b185c5ba097)

If PTE has valid and non-reserved RWX permissions, but the translated Physical address (pte.ppn of leaf PTE + offset) does not have (r,w,x) PMP permissions, then accessing the translated Physical address would raise access fault exception.
When satp.mode=sv32, and PTE has valid and non-reserved RWX permissions then test the following in supervisor and user privilege mode for level1 and level0 PTE.

- Remove read PMP permission of translated Physical Address in pmpcfg and test the read acces.
- Remvoe write PMP permission of translated Physical Address in pmpcfg and test the write access.
- Remove execute PMP permission of translated Physical Address in pmpcfg and test the execute access.

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_pmp_check-01.S** 	|   Supervisor	|     0     	|    Execute Access  	|
| **vm_pmp_check-02.S** 	|   Supervisor	|     0     	|   Read  Access      |   
| **vm_pmp_check-03.S** 	|   Supervisor 	|     0     	|   Write Access  |   
| **vm_pmp_check-04.S** 	|   Supervisor 	|     1     	|   Execute Access 	|
| **vm_pmp_check-05.S** 	|   Supervisor 	|     1     	|   Read  Access      | 
| **vm_pmp_check-06.S** 	|   Supervisor 	|     1     	|  Write Access    | 
| **vm_pmp_check-07.S** 	|   User	|     0     	|   Execute Access  	|
| **vm_pmp_check-08.S** 	|   User	|     0     	|  Read  Acces     |   
| **vm_pmp_check-09.S** 	|   User 	|     0     	|   Write Access  |   
| **vm_pmp_check-10.S** 	|   User 	|     1     	|  Execute  Access  	|
| **vm_pmp_check-11.S** 	|   User 	|     1     	|   Read  Access       | 
| **vm_pmp_check-12.S** 	|   User 	|     1     	|  Write Access   |

Ammar-10xe committed:
Access bit implementation

Ammar-10xe commented:
## [Test for Access bit implementation](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/07441863de8026a1c79557f33b8e38a5149ab5bb)

If implementation does not sets the pte.A on accessing the PTE, and PTE has pte.A=0, then accessing it would raise page fault exception of the corresponding access type. 
When satp.mode=sv32, PTE has (r,w,x) PMP permissions, and pte.v=1, then test the following in supervisor and user privilege mode for level0 and level1 PTE. Execute sfence.vma before accessign the PTE.

- Set pte.r=1 & pte.a=0 and test the read acces. 
- Set pte.w=1 & pte.a=0 and test the write access.
- Set pte.x=1 & pte.a=0 and test the execute access.

**NOTE:** For testing at level0, set all the pte perimssions at level1 to 0 except pte.v, so that level1 points to level0. Set pte.U=0 when test in Supervisor mode and Set pte.U=1 when testing in user mode.

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_access_bit-01.S** 	|   Supervisor	|     0     	|    Execute Access  	|
| **vm_access_bit-02.S** 	|   Supervisor	|     0     	|   Read  Access      |   
| **vm_access_bit-03.S** 	|   Supervisor 	|     0     	|   Write Access  |   
| **vm_access_bit-04.S** 	|   Supervisor 	|     1     	|   Execute Access 	|
| **vm_access_bit-05.S** 	|   Supervisor 	|     1     	|   Read  Access      | 
| **vm_access_bit-06.S** 	|   Supervisor 	|     1     	|  Write Access    | 
| **vm_access_bit-07.S** 	|   User	|     0     	|   Execute Access  	|
| **vm_access_bit-08.S** 	|   User	|     0     	|  Read  Acces     |   
| **vm_access_bit-09.S** 	|   User 	|     0     	|   Write Access  |   
| **vm_access_bit-10.S** 	|   User 	|     1     	|  Execute  Access  	|
| **vm_access_bit-11.S** 	|   User 	|     1     	|   Read  Access       | 
| **vm_access_bit-12.S** 	|   User 	|     1     	|  Write Access   |

Ammar-10xe committed:
Make Executable Page Readable with s/mstatus.MXR unset

Ammar-10xe commented:
## [Test for Make Executable Page Readable with s/mstatus.MXR unset](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/779ade5674b35bb38da269d2e0ce85ae0c17d9ff)

If PTE has only execute permission (pte.x = 1) and s/mstatus.MXR=0, then read access on that PTE should raise load page fault exception.
When satp.mode=sv32, PTE has (r,w,x) PMP permissions, and pte.v=1, then test the following in supervisor and user privilege mode for level0 and level1 PTE.
- Set pte.r=0 & pte.w=0 & pte.x=1 & s/mstatus.MXR=0 and test the read acces.

 For testing at level0, set all the pte perimssions at level1 to 0 except pte.v, so that level1 points to level0.
 Set pte.U=0 when test in Supervisor mode and Set pte.U=1 when testing in user mode

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_mxr_unset-01.S** 	|   Supervisor	|     0     	|   Read Access  	|
| **vm_mxr_unset-02.S** 	|   Supervisor	|     1     	|   Read Access      |   
| **vm_mxr_unset-03.S** 	|   User	|     0    	|   Read Access |   
| **vm_mxr_unset-04.S** 	|   User 	|     1     	|   Read Access  	|

Commit:
Test for No Dirty Bit Implementation
If implementation does not sets the pte.D when PTE is written,
and PTE has pte.D=0, then attempting to store on that PTE would
raise Store/AMO page fault exception.

Ammar-10xe committed:
tests discription added

Ammar-10xe committed:
Merge branch 'sv32-arch-test' of github.com:10x-Engineers/core-v-verif into sv32-arch-test

MuhammadHammad001 commented:
## [Test for No Dirty Bit Implementation](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/f1d4daf09cd074bc2f4f9768c128953b73a6408b)

#### Feature Description:
If implementation does not sets the pte.D when PTE is written, and PTE has pte.D=0, then attempting to store on that PTE would raise **Store/AMO page fault exception**.

#### Files and their purpose:

|  **`File Name`**  | **`Mode`** | **`Level`** | **`Test Access`** |    **`Expected`**    |
|:-----------------:|:----------:|:-----------:|:-----------------:|:--------------------:|
| **vmDirtyBit-01** | Supervisor |      1      |     Write Test    | Store/AMO page fault |
| **vmDirtyBit-02** |    User    |      1      |     Write Test    | Store/AMO page fault |
| **vmDirtyBit-03** | Supervisor |      0      |     Write Test    | Store/AMO page fault |
| **vmDirtyBit-04** |    User    |      0      |     Write Test    | Store/AMO page fault |

Ammar-10xe committed:
vm_pte_u_set and vm_pte_u_unset test name changed

Commit:
Tests for Reserved RWX Permission encodings of PTE
If PTE has reserved RWX encodings (pte.w=1 & pte.r=0),
then accessing it would raise page fault exception of
the corresponding access type.

HAMZA-AFZAL404 committed:
pmp on pte tests added

Commit:
Merge branch 'sv32-arch-test' of https://github.com/10x-Engineers/core-v-verif into sv32-arch-test

HAMZA-AFZAL404 commented:
## [Test for PMP permissions on PTE](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/01b47da9135b7b05b2ac450a49ba18e79dd4e22b)

If PTE does not have (r,w,x) PMP permissions, then accessing it would raise access fault exception of the corresponding access type. 
When satp.mode=sv32, then test the following in supervisor and user privilege mode for level0 and level1 PTE.
- Remove read PMP permission of PTE address in pmpcfg and test the read acces.
- Remvoe write PMP permission of PTE address in pmpcfg and test the write access.
- Remove execute PMP permission of PTE address in pmpcfg and test the execute access.

**NOTE:** For testing at level0, set all the pte perimssions at level1 to 0 except pte.v, so that level1 points to level0. Set pte.U=0 when test in Supervisor mode and Set pte.U=1 when testing in user mode.

#### Files and their purpose:
|   **`File Name`**   	|  **`Mode`**  	| **`Level`** 	| **`Testing For`** 	| 
|:------------------------:|:------------------:|:-----------------:|:-------------------------:|
| **vm_access_bit-01.S** 	|   Supervisor	|     0     	|    Execute Access  	|
| **vm_access_bit-02.S** 	|   Supervisor	|     0     	|   Read  Access      |   
| **vm_access_bit-03.S** 	|   Supervisor 	|     0     	|   Write Access  |   
| **vm_access_bit-04.S** 	|   Supervisor 	|     1     	|   Execute Access 	|
| **vm_access_bit-05.S** 	|   Supervisor 	|     1     	|   Read  Access      | 
| **vm_access_bit-06.S** 	|   Supervisor 	|     1     	|  Write Access    | 
| **vm_access_bit-07.S** 	|   User	|     0     	|   Execute Access  	|
| **vm_access_bit-08.S** 	|   User	|     0     	|  Read  Acces     |   
| **vm_access_bit-09.S** 	|   User 	|     0     	|   Write Access  |   
| **vm_access_bit-10.S** 	|   User 	|     1     	|  Execute  Access  	|
| **vm_access_bit-11.S** 	|   User 	|     1     	|   Read  Access       | 
| **vm_access_bit-12.S** 	|   User 	|     1     	|  Write Access   |

MuhammadHammad001 commented:
## [Test for Reserved RWX Permission encodings of PTE](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/5062b8a06177902020bd865d1450964895adad7c)

#### Feature Description:
If PTE has reserved RWX encodings (pte.w=1 & pte.r=0 & pte.x=0), then accessing it would raise page fault exception of the corresponding access type. 

#### Files and their purpose:
|    **`File Name`**   | **`Mode`** | **`Level`** | **`Test Access`** |        **`Expected`**        |
|:--------------------:|:----------:|:-----------:|:-----------------:|:----------------------------:|
| **vmRWXreserved-01** | Supervisor |      1      |     Read Test     |        Load Page Fault       |
| **vmRWXreserved-02** | Supervisor |      1      |     Write Test    |     Store/AMO page fault     |
| **vmRWXreserved-03** | Supervisor |      1      |    Execute Test   | Fetch/Instruction page fault |
| **vmRWXreserved-04** |    User    |      1      |     Read Test     |        Load Page Fault       |
| **vmRWXreserved-05** |    User    |      1      |     Write Test    |     Store/AMO page fault     |
| **vmRWXreserved-06** |    User    |      1      |    Execute Test   | Fetch/Instruction page fault |
| **vmRWXreserved-07** | Supervisor |      0      |     Read Test     |        Load Page Fault       |
| **vmRWXreserved-08** | Supervisor |      0      |     Write Test    |     Store/AMO page fault     |
| **vmRWXreserved-09** | Supervisor |      0      |    Execute Test   | Fetch/Instruction page fault |
| **vmRWXreserved-10** |    User    |      0      |     Read Test     |        Load Page Fault       |
| **vmRWXreserved-11** |    User    |      0      |     Write Test    |     Store/AMO page fault     |
| **vmRWXreserved-12** |    User    |      0      |    Execute Test   | Fetch/Instruction page fault |

Ammar-10xe committed:
PTE invalid test 06 updated

Ammar-10xe committed:
vm_satp_access test updated

Commit:
Test for Non-leaf PTE permission at level 0
If PTE at level0 has non-leaf RWX permissions (pte.x=0 & pte.r=0 & pte.w=0),
then accessing it would raise page fault exception of the corresponding
access type.

MuhammadHammad001 commented:
## [Test for Non Leaf PTE permissions at Level 0](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/5fbb6247a5f2f7c8d336a84171664ce77deadb68)
### Feature Description
If PTE at level0 has non-leaf RWX permissions (pte.x=0 & pte.r=0 & pte.w=0), then accessing it would raise page fault exception of the corresponding access type.

### Files and their purpose
|  **`File Name`**  | **`Mode`** | **`Level`** | **`Test Access`** |        **`Expected`**        |
|:-----------------:|:----------:|:-----------:|:-----------------:|:----------------------------:|
| **vm_NLeaf_PTE_01** | Supervisor |      0      |     Read Test     |        Load Page Fault       |
| **vm_NLeaf_PTE_02** | Supervisor |      0      |     Write Test    |     Store/AMO page fault     |
| **vm_NLeaf_PTE_03** | Supervisor |      0      |    Execute Test   | Fetch/Instruction page fault |
| **vm_NLeaf_PTE_04** |    User    |      0      |     Read Test     |        Load Page Fault       |
| **vm_NLeaf_PTE_05** |    User    |      0      |     Write Test    |     Store/AMO page fault     |
| **vm_NLeaf_PTE_06** |    User    |      0      |    Execute Test   | Fetch/Instruction page fault |

Commit:
Non Leaf PTE test files renamed

Commit:
Test for mstatus.tvm
If mstatus.TVM=1, read and write access to the satp and sfence.vma
will raise illegal instruction exception in S-mode.

MuhammadHammad001 commented:
## [Test for mstatus.tvm](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/e050043162367b3bcae6ea703dcffc04fa512fd8)
### Feature Descritption
If mstatus.TVM=1, read and write access to the satp and sfence.vma will raise illegal instruction exception in S-mode.
### Files and their purpose
|   **`File Name`**  | **`Mode`** | **`Level`** |    **`Test Access`**   |       **`Expected`**      |
|:------------------:|:----------:|:-----------:|:----------------------:|:-------------------------:|
| **vm_mstatus_tvm_01** | Supervisor |      1      |    Test      -- sfence | Illegal Instruction fault |
| **vm_mstatus_tvm_02** | Supervisor |      1      |   Read Test  -- satp   | Illegal Instruction fault |
| **vm_mstatus_tvm_03** | Supervisor |      1      |   Write Test -- satp   | Illegal Instruction fault |
| **vm_mstatus_tvm_04** | Supervisor |      0      |    Test      -- sfence | Illegal Instruction fault |
| **vm_mstatus_tvm_05** | Supervisor |      0      |   Read Test  -- satp   | Illegal Instruction fault |
| **vm_mstatus_tvm_06** | Supervisor |      0      |   Write Test -- satp   | Illegal Instruction fault |

Commit:
Description update for mstatus.tvm tests

Commit:
Test for Satp Mode Field set
Verification of supported address translation scheme i.e sv32
is selected by writing satp.mode=sv32 and reading back the satp

Commit:
Description update for satp mode set test

MuhammadHammad001 commented:
## [Test for satp mode Field Set](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/60bb2d32cc3d9fa4664e45163165b69a771defce)

### Feature Description
Verification of supported address translation scheme i.e sv32
is selected by writing satp.mode=sv32 and reading back the satp

### Files and their purpose
|    **`File Name`**   | **`Mode`** | **`Level`** |      **`Test Access`**     |         **`Expected`**        |
|:--------------------:|:----------:|:-----------:|:--------------------------:|:-----------------------------:|
| **vm_satp_mode_set** | Supervisor |      1      |   Write and then read satp | Pass with satp mode field = 1 |

Commit:
Test for Make Executable Page Readable with s/mstatus.MXR set added
If PTE has only execute permission (pte.x = 1) and s/mstatus.MXR=1,
then read access on that PTE should be successful without having
explicit read permission (pte.r=0).

Commit:
MXR bit set test files renamed

MuhammadHammad001 commented:
## [Test for Make Executable page Readable with mstatus.mxr set](https://github.com/10x-Engineers/core-v-verif/pull/9/commits/bbed174bfd6828abc77d9eaa5d6877e8ccd8c2d0)

### Feature Description
If PTE has only execute permission (pte.x = 1) and s/mstatus.MXR=1,
then read access on that PTE should be successful without having
explicit read permission (pte.r=0).

### Files and their purpose
|  **`File Name`**  | **`Mode`** | **`Level`** | **`Test Access`** | **`Expected`** |
|:-----------------:|:----------:|:-----------:|:-----------------:|:--------------:|
| **vm_MXR_set_01** | Supervisor |      1      |     Read Test     |      Pass      |
| **vm_MXR_set_02** |    User    |      1      |     Read Test     |      Pass      |
| **vm_MXR_set_03** | Supervisor |      0      |     Read Test     |      Pass      |
| **vm_MXR_set_04** |    User    |      0      |     Read Test     |      Pass      |
